# MetaGaia: A tool for analyzing bin abundances and metabolic profiles

<img src="https://valdeanda.github.io/MetaGaia/docs/Metagaia_summ.png" width="200" height="200" align="right">

Last updated:  May 19th, 2021

### `Requirements`



---

### `About Metagaia`
MetaGaia is a pipeline that streamlines the process of calculating Metagenome-Assambled-Genomes (MAGs) relative abundance in metagenomic communities, identifying the metabolic pathways present, and the identification of Auxiliary Metabolic Genes (AMGs) present in in both viral phages and their MAGs hosts. 
The importance of performing each of these functions is to better understand microbial communities and the dominant microbes present within a community.



This pipeline is divided into three sections that perform different functions: 

1. Calculating bin abundances (abundance)
2. Identifying metabolic pathways present in each bin (metabolism)
3. Finding commonalities between phages and their hosts (host-virus).


The Gaia hypothesis, also known as Gaia theory or Gaia principle, proposes that all organisms and their inorganic surroundings on Earth are closely integrated to form a single and self-regulating complex system, maintaining the conditions for life on the planet

[Lovelock, J.E and Margulis, L. (1974). “Atmospheric homeostasis by and for the biosphere- The Gaia hypothesis”. Tellus 26 (1): 2–10](https://onlinelibrary.wiley.com/doi/abs/10.1111/j.2153-3490.1974.tb01946.x)

--- 

**### `Readme`**
View the full documentation [here](https://valdeanda.github.io/MetaGaia/docs/READMEv1.html).

---
**### `Citation`**
If you find MetaGaia useful for your research, please cite the biorXiv preprint as:

Coming soon!

---


### `Developers`

* Sahil B. Shah - sahilbshah@berkeley.edu, 
* Marguerite Langwig - marguerite.langwig@utexas.edu, 
* Valerie De Anda - valdeanda@utexas.edu*



### `Developers`


